KEP-4800: Split UnCoreCache awareness (#4810)
kubernetes/enhancements: keps/sig-node/4800-cpumanager-split-uncorecache/README.md

- Topology struct already contains CPUDetails which is a map of CPUInfo.  CPUInfo knows about NUMA, socket, and core IDs associated with a CPU.  We just need to add a new member called uncorecacheId into the CPUInfo struct that tracks whether the CPU is part of a split uncore caches. Functionality to support this was merged in cadvisor with these pull-requests [cadvisor/pr-2849](https://github.com/google/cadvisor/pull/2849) and [cadvisor/pr-2847](https://github.com/google/cadvisor/pull/2847/)

- Handle enablement of the policy option in pkg/kubelet/cm/cpumanager/policy_options.go and check the validity of the user flag and capability of the system.  If topology does not support split uncore caches, grouping by uncore cache static policy will be ignored.  

- Modify the "Allocate" static policy to check for the option, prefer-align-cpus-by-uncorecache. For platforms where SMT is enabled, prefer-align-cpus-by-uncorecache will continue to follow default behaviour and try to allocate full cores when possible. prefer-align-cpus-by-uncorecache can be enabled along with full-pcpus-only and enforce full core assignment with uncorecache alignment.

> - prefer-align-cpus-by-uncorecache will be compatible with the default CPU allocation logic with future plans to be compatible with the options distribute-cpus-across-numa and distribute-cpus-across-cores.

Reviewer: Does this mean that for now you cannot combine these options together? What happens if you do?

Reviewer 2: Correct, for now when the mentioned flags are enabled will cause a failure and node will not be schedulable.

Reviewer 3: We reached agreement that expanding/improving compatibility between options is a Beta graduation blocker.

### Notes/Constraints/Caveats (Optional)

- The name UnCoreCache is directly derived from cAdvisor that is being used as a package. 
- Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/af879aebb1a866a2f0e45bb33c09a1cc8f7acc45/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

### Risks and Mitigations

- Risk: Enabling or disabling the feature might lead to unexpected behavior
  - Mitigation: Feature is enabled by a static policy option flag
- Risk: The new feature might interfere with existing functionality
  - Mitigation: It does not change behavior of non-static policy and preserves the behavior of other static options
- Risk: Inconsistent configuration could cause scheduling issues

>   - Mitigation: Failure will occur if mismatch between options occurs, preventing the Pod from being scheduled incorrectly or leading a non-optimal aligment

Reviewer 4: Can we clarify where the failure occurs? As written it suggests: 
1. We prevent incompatible options from being specifie and fail on kubelet config validation. 
OR
2. We fail at runtime when trying to place a pod when incompatible options are specified which is _not_ desirable.

Author: We are suggesting the 2nd option that we would fail at runtime

Author: I went ahead and added a small change, does that clear it up a bit?

The algorithm for prefer-align-cpus-by-uncorecache will be implemented to follow the default packed behavior of the existing static CPUManager allocation with the introduction of a uncorecache hierarchy. This means that when a guaranteed container requires CPUs that are equal to or greater than the size of a NUMA or socket, the CPU allocation will behave as usual and schedule the full socket or NUMA. The scheduled CPUs will be subtracted from the quantity of CPUs required for the container.

> When/once the required CPUs for a container are less than the number of CPUs within a NUMA, the algorithm will be implemented as follows:

Reviewer: ```suggestion
When/once the required CPUs for a container are less than the number of CPUs within a NUMA node, the algorithm will be implemented as follows:
```

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Author: I checked off whatever I knew for sure but some checks I wasn't sure about

If the cpumanager cannot align optimally, it will still admit the workload as before. Uncore cache assignment will be preferred but not a requirement for this feature.
This feature is does not introduce the requirement of aligned by uncorecache but the preference to alignment of uncorecache.

## Motivation

Main motivation for this proposal is to reduce noisy neighbor scenarios that occur on systems with split uncore cache, which is available on both x86 and ARM architecture.
The challenge with current kubelet’s cpu manager is that it is unaware of split uncore architecture and spreads CPU assignments across distributed uncore caches. This creates a noisy neighbor problem where multiple pods/containers are sharing the same uncore cache. In addition, pods spread against multiple uncore caches see higher latency and reduced performance due to inter-cache access latency. For workload use cases that are sensitive to latency and are performance deterministic, minimizing the noisy neighbor condition in uncore cache can have significant improvements in performance.

> ![Figure to the motivation](https://github.com/ajcaldelas/enhancements/blob/split_l3_cache/keps/sig-node/4800-cpumanager-split-uncorecache/images/uncore-alignment-motivation.png)

Author: Oh nice thank you!

#### Story 2

- As a Networking/Telco Engineer, my application is sensitive to uncore cache latency. I want my application to be statically assigned to CPUs that correspond to one single uncore cache so that I can get the maximum throughput from my application. Additionally, I have applications that are also static and of different CPU size requirements. The default Static CPU Manager Policy will result in my applications being assigned to two or more uncore caches. I have a large multi-node Kuberentes cluster. I want a scheduling policy to automatically assign CPUs for the best performance across the fewest amount of uncore cahces based on available CPUs without having to worry about what order/size I deploy my applications so I can ensure I get the best performance possible. 

### Notes/Constraints/Caveats (Optional)

- The name UnCoreCache is directly derived from cAdvisor that is being used as a package. 

> - Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

Author: Adding this right? 
https://github.com/kubernetes/kubernetes/blob/af879aebb1a866a2f0e45bb33c09a1cc8f7acc45/pkg/features/kube_features.go#L110C1-L117C83

> - Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

Author: I think I did this right, so Ill resolve it for now

Feature will be enabled. Proper drain of node and restart of kubelet required. Feature is not intended to be enabled/disabled dynamically, similar to static policy.

###### Are there any tests for feature enablement/disablement?

Option is not enabled dynamically. To enable/disable option, cpu_manager_state must be removed and kubelet must be restarted.
Unit tests will be implemented to test if the feature is enabled/disabled.
E2e node serial suite can be use to test the enablement/disablement of the feature since it allows the kubelet to be restarted.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

Kubelet restarts are not expected to impact existing CPU assignments to already running workloads

###### What specific metrics should inform a rollback?

N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

> N/A

Reviewer 5: ```suggestion
No
```

Reference podresources API to determine CPU assignment and CacheID assignment.

###### How can someone using this feature know that it is working for their instance?

Reference podresources API to determine CPU assignment and CacheID assignment.

<!--
For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
for each individual pod.
Pick one more of these and delete the rest.
Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

Measure the time to deploy pods under default settings and compare to the time to deploy pods with align-by-uncorecache enabled. Time difference should be negligible.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- Metrics

>   - Pod deployment time (seconds)

Reviewer 5: Can you point out specific metric here?

Author: We don't specifically have a known metric to measure this, that's part of our proposal to add a new metric would we mention that here?

Reviewer 3: For this and for https://github.com/kubernetes/enhancements/pull/4810#discussion_r1784594941 we can use `topology_manager_admission_duration_ms`  - topology manager since GA will _always_ be engaged even if no NUMA alignment is requested, so this is a good approximation. I'd be open to add more metrics in the future, but this can be a nice start and perhaps enough for alpha.

Author: okay I added something

      - [Unit tests](#unit-tests)
      - [Integration tests](#integration-tests)
      - [e2e tests](#e2e-tests)
  - [Graduation Criteria](#graduation-criteria)
    - [Alpha](#alpha)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
  - [Monitoring Requirements](#monitoring-requirements)
  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)

> - [ ] (R) KEP approvers have approved the KEP status as `implementable`

Reviewer 3: IIRC we can check this because we are considering this for 1.32 right @haircommander @SergeyKanzhelev ?

Reviewer 6: yeah it should be marked as X (though it's not yet been approved, so I understand the confusion and it's a weird interaction)

- [x] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 

> - [x] (R) Production readiness review completed

Reviewer 3: not yet done, but you can leave checked, if it merges PRR is approved

Author: Okay, Ill add this. Besides this, was my assumptions for the other checks correct?

#### Story 2

- As a Networking/Telco Engineer, my application is latency sensitive, to which uncore cache misalignment is a contributing factor. I want my application to be statically assigned to CPUs that correspond to one single uncore cache so that I can get the maximum throughput from my application. Additionally, I have applications that are also static and of different CPU size requirements. The default Static CPU Manager Policy will result in my applications being assigned to two or more uncore caches. I have a large multi-node Kuberentes cluster. I want a cpu allocation logic to automatically assign CPUs for the best performance across the fewest amount of uncore cahces based on available CPUs without having to worry about what order/size I deploy my applications so I can ensure I get the best performance possible. 

### Notes/Constraints/Caveats (Optional)

> - The name UnCoreCache is directly derived from cAdvisor that is being used as a package. 

Reviewer 3: nit: let's make a pass to make sure we are consistent with either `UnCoreCache` or `UncoreCache` here in the document and in the future also in the codebase. I'm fine with either wordings.

Author: Okay let's just do UncoreCache

> - Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/af879aebb1a866a2f0e45bb33c09a1cc8f7acc45/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

Reviewer 4: ```suggestion
- Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/af879aebb1a866a2f0e45bb33c09a1cc8f7acc45/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid proliferation of feature gates. 
```

When/once the required CPUs for a container are less than the number of CPUs within a NUMA, the algorithm will be implemented as follows:
1. Scan each socket in numerical order. If required CPUs are less than the size of available CPUs on the socket, pick this socket of CPUs.
2. Within the chosen socket, scan each NUMA in numerical order. If the required CPUs are less than the number of the available CPUs on the NUMA, pick the subset of CPUs that correspond to this NUMA.
3. Within the chosen NUMA, scan through every uncore cache index in numerical order:
   - If the required CPUs are greater than or equal the size of the **total** amount of CPUs to an uncore cache, assign the full uncore cache CPUs if they are all **unscheduled** within the group. Subtract the amount of scheduled CPUs from the quantity of required CPUs. 
     - Continue scanning uncore caches in numerical order and assigning full uncore cache CPU groups until the required quantity is less than the **total** number of CPUs to a uncore cache group.

>    - If the required CPUs are less than the size of the **total** amount of CPUs to an uncore cache, scan through each uncore cache index in numerical order starting from the first index, and assign CPUs if there are enough **available** CPUs within the uncore cache group.

Reviewer 3: asking to confirm one final time my understanding and the intent: this means we tend to pack uncore groups and we tend to fully utilize uncore groups before to move to others, right?

Reviewer 2: correct, pack only if the container can still fit within the uncore CPU group, otherwise move to the next uncore

Author: Yes, I would say yes that's what we're doing.
@wongchar @sphrasavath just to get some additional eyes on this

Reviewer 3: perfect, thanks. No further comments from me atm.

  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> We proposed this KEP to introduce a new CPU Manager static policy option, "align-cpus-by-uncorecache", that groups CPU resources by uncore cache where possible.

Reviewer 4: ```suggestion
This KEP proposes to introduce a new CPU Manager static policy option, "align-cpus-by-uncorecache", that groups CPU resources by uncore cache where possible.
```

The opt-in feature changes the cpu assignment algorithm to add sorting by uncore cache and then taking cpus aligned to the same uncore cache, where possible.
In cases where numbers of cpu requested exceeds number of cpus grouped in the same uncore cache, the algorithm attempts best-effort to reduce assignments of cpus across minimal numbers of uncore caches.

> If the cpumanager can’t align optimally, it will still admit the workload as before. Uncore cache assignment will be preferred but not a requirement for this feature. More details will be given in the motivation section.

Reviewer 4: ```suggestion
If the cpumanager can’t align optimally, it will still admit the workload as before. Uncore cache assignment will be preferred but not a requirement for this feature.
```

## Motivation

> Our motivation is to reduce noisy neighbor scenarios that occur on systems with split uncore cache, which is available on both x86 and ARM architecture.

Reviewer 4: ```suggestion
Main motivation for this proposal is to reduce noisy neighbor scenarios that occur on systems with split uncore cache, which is available on both x86 and ARM architecture.
```

> The challenge with current kubelet’s cpu manager is that it is unaware of split uncore architecture and spreads CPU assignments across distributed uncore caches. This creates a noisy neighbor problem where multiple pods/containers are sharing the same uncore cache. In addition, pods spread against multiple uncore caches sees higher latency and reduced performance due to inter-cache access latency. For workload use cases that are sensitive to latency and are performance deterministic, minimizing the noisy neighbor condition in uncore cache can have significant improvements in performance.

Reviewer 4: ```suggestion
The challenge with current kubelet’s cpu manager is that it is unaware of split uncore architecture and spreads CPU assignments across distributed uncore caches. This creates a noisy neighbor problem where multiple pods/containers are sharing the same uncore cache. In addition, pods spread against multiple uncore caches see higher latency and reduced performance due to inter-cache access latency. For workload use cases that are sensitive to latency and are performance deterministic, minimizing the noisy neighbor condition in uncore cache can have significant improvements in performance.
```

### Goals

- Introduce a new CPU Manager policy option that assigns CPU within the same grouping of uncore cache to container scope.  
- Minimize the number of CPUs assignments from different uncore cache grouping to pods and container scope.

### Non-Goals

- This proposal does not aim to modify CPU assignments for CPU Manager policy set to none.
- This does not alter the behavior of existing static policy options such as full-pcpus-only.

## Proposal

- Add a new static policy option to fine-tune of CPU assignments to align by uncore cache grouping.

- Topology struct already contains CPUDetails which is a map of CPUInfo.  CPUInfo knows about NUMA, socket, and core IDs associated with a CPU.  We just need to add a new member called uncorecacheId into the CPUInfo struct that tracks whether the CPU is part of a split uncore caches. Functionality to support this was merged in cadvisor with these pull-requests [cadvisor/pr-2849](https://github.com/google/cadvisor/pull/2849) and [cadvisor/pr-2847](https://github.com/google/cadvisor/pull/2847/)

> - Handle enablement of the policy option in pkg/kubelet/cm/cpumanager/policy_options.go and check the validity of the user flag and capability of the system.  If topology does not support split uncore caches, grouping by uncore cache static policy will be ignored.  

Reviewer 4: Do we warn when we are unable to align? Should the name be prefixed with `prefer` so in the future we may consider hard failing when conditions aren't met?

Reviewer 2: The implementation is preferred so in the case where the node is highly utilized and the available CPUs are fractured across multiple uncore caches, the container will still be scheduled as long as the CPU limit can be met. We can add the prefix to make that more clear.

Reviewer 3: I'd be slightly in favor of renaming `prefer-align-cpus-by-uncorecache` or any shorter name which conveys the intent.

- Modify the "Allocate" static policy to check for the option, align-cpus-by-uncorecache. For platforms where SMT is enabled, align-cpus-by-uncorecache will continue to follow default behaviour and try to allocate full cores when possible. Align-cpus-by-uncorecache can be enabled along with full-pcpus-only and enforce full core assignment with uncorecache alignment.

> - Align-cpus-by-uncorecache will be compatible with the default scheduling algo with future plans to be compatible with the options distribute-cpus-across-numa and distribute-cpus-across-cores.

Reviewer 4: ```suggestion
- Align-cpus-by-uncorecache will be compatible with the default scheduling algorithm with future plans to be compatible with the options distribute-cpus-across-numa and distribute-cpus-across-cores.
```
```suggestion
- Align-cpus-by-uncorecache will be compatible with the default scheduling algo with future plans to be compatible with the options distribute-cpus-across-numa and distribute-cpus-across-cores.
```

### Notes/Constraints/Caveats (Optional)

- The name UnCoreCache is directly derived from cAdvisor that is being used as a package. 
- Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

### Risks and Mitigations

- Feature is enabled by a static policy option flag.  
- It does not change behavior of non-static policy

> - It preserves the behavior of other static options.

Reviewer 4: Will we fail if we detect options that aren't mutually compatible?

Author: Yes, it will fail if we aren't using compatible it should tell the user that we are not able to schedule the pod
Ill add verbiage here to reflect it.

Author: Added some wording about it.

## Design Details

To make Kubelet be uncore cache aware for CPU sensitive workloads the CPU allocation, `UnCoreCacheID` has been added to the `CPUInfo` and `CPUTopology` structures that are provided by `cAdvisor` during the CPU discovery process.

> A new CPUManager policy option will be introduced, `align-cpus-by-uncorecache`, to enable this feature. With this feature enabled, a modification of the existing default scheduling function is needed to take into consideration of the UnCoreCache. The allocation process will first require to attempt to allocate CPUs of a full socket. Then allocation should attempt to allocate the CPUs within the logical boundary of a NUMA node within the socket. Then allocation then must consider the subset of CPUs that correspond to each UnCoreCache within the prior subset of CPUs. Uncore cache allocation will be preferred but not a strict requirement. Since this policy augments the CPUManager's Static policy, this feature will only be applied to guaranteed pods with whole CPU requests.

Reviewer 4: ```suggestion
A new CPUManager policy option, align-cpus-by-uncorecache, will be introduced to enable this feature. When enabled, the existing default scheduling function will be modified to account for UnCoreCache. The allocation process will first attempt to allocate CPUs from an entire socket. Next, it will attempt to allocate CPUs within the logical boundary of a NUMA node within the socket. Finally, the allocation will consider the subset of CPUs associated with each UnCoreCache within the previously selected CPUs. While UnCoreCache allocation is preferred, it is not strictly required. Since this policy extends the CPUManager’s Static policy, it will only apply to guaranteed pods with whole CPU requests.
```

  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This KEP proposes to introduce a new CPU Manager static policy option, "align-cpus-by-uncorecache", that groups CPU resources by uncore cache where possible.

Reviewer 3: IIUC we want to call this `prefer-align-cpus-by-uncorecache`, right? If so let's fix all the instance of the name.

Author: Okay Ill make the change

- Topology struct already contains CPUDetails which is a map of CPUInfo.  CPUInfo knows about NUMA, socket, and core IDs associated with a CPU.  We just need to add a new member called uncorecacheId into the CPUInfo struct that tracks whether the CPU is part of a split uncore caches. Functionality to support this was merged in cadvisor with these pull-requests [cadvisor/pr-2849](https://github.com/google/cadvisor/pull/2849) and [cadvisor/pr-2847](https://github.com/google/cadvisor/pull/2847/)

- Handle enablement of the policy option in pkg/kubelet/cm/cpumanager/policy_options.go and check the validity of the user flag and capability of the system.  If topology does not support split uncore caches, grouping by uncore cache static policy will be ignored.  

- Modify the "Allocate" static policy to check for the option, align-cpus-by-uncorecache. For platforms where SMT is enabled, align-cpus-by-uncorecache will continue to follow default behaviour and try to allocate full cores when possible. Align-cpus-by-uncorecache can be enabled along with full-pcpus-only and enforce full core assignment with uncorecache alignment.

> - Align-cpus-by-uncorecache will be compatible with the default scheduling algorithm with future plans to be compatible with the options distribute-cpus-across-numa and distribute-cpus-across-cores.

Reviewer 3: perhaps it's me, but "scheduling algorithm" makes me thing about kube-scheduler more than the cpu allocation logic. So I'd rephrase like "the default cpu allocation logic"

Author: Added

### User Stories (Optional)

#### Story 1

- As a HPC user, I want to extract the best performance possible and reduce latency, so that I can extract the most value when deploying applications in Kubernetes. I can use the static CPU Manager Policy option, but on split uncore cache processors, my application can experience latency when assigned CPUs across multiple uncore caches. In order to maximize performance, I want to minimize the distribution across uncore caches to minimize the latency. I want the scheduler to automatically handle this for me for larger clusters for simple deployment and efficiency.

> 

Reviewer 3: in this context the "scheduler" is the kube-scheduler or the cpu allocation logic inside the kubelet?

Author: The cpu allocation logic inside the kubelet

Reviewer 2: updated

#### Story 2

- As a Networking/Telco Engineer, my application is sensitive to uncore cache latency. I want my application to be statically assigned to CPUs that correspond to one single uncore cache so that I can get the maximum throughput from my application. Additionally, I have applications that are also static and of different CPU size requirements. The default Static CPU Manager Policy will result in my applications being assigned to two or more uncore caches. I have a large multi-node Kuberentes cluster. I want a scheduling policy to automatically assign CPUs for the best performance across the fewest amount of uncore cahces based on available CPUs without having to worry about what order/size I deploy my applications so I can ensure I get the best performance possible. 

> 

Reviewer 3: I'm fairly confident telco application are latency sensitive in general, and uncore cache misalignment can be a contributing factor of that latency, which can be actually non-obvious when sub-numa clustering (or equivalent) is disabled so the CPU present itself as a single NUMA node.

So, I'd rephrase as "my application is latency sensitive, to which uncore cache misalignment is a contributing factor"

### Notes/Constraints/Caveats (Optional)

- The name UnCoreCache is directly derived from cAdvisor that is being used as a package. 
- Using the flow in `CPUManagerPolicy{Alpha,Beta}Options` from [xref](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L110C1-L117C83) which is used to avoid a proliferation of a feature gates. 

### Risks and Mitigations

- Feature is enabled by a static policy option flag.  
- It does not change behavior of non-static policy
- It preserves the behavior of other static options.

> - Failure will occur if mismatch between options occurs and the Pod will not be scheduled.

Reviewer 3: this is unclear though. A risk factor could be that since the feature is best-effort, low-latency workload can be admitted without ensuring optimal uncore cache alignment; but we will be adding metrics as mitigation factor so this can be detected.

Another obvious risk is bugs make the kubelet crash; a mitigation factor is twofold: we can disable the feature using the cpumanager policy option, and while non-GA we can disable with the feature gate.

## Design Details

To make Kubelet be uncore cache aware for CPU sensitive workloads the CPU allocation, `UnCoreCacheID` has been added to the `CPUInfo` and `CPUTopology` structures that are provided by `cAdvisor` during the CPU discovery process.

> A new CPUManager policy option, align-cpus-by-uncorecache, will be introduced to enable this feature. When enabled, the existing default scheduling function will be modified to account for UnCoreCache as a preference. The allocation process will first attempt to allocate CPUs from an entire socket. Next, it will attempt to allocate CPUs within the logical boundary of a NUMA node within the socket. Finally, the allocation will consider the subset of CPUs associated with each UnCoreCache within the previously selected CPUs. While UnCoreCache allocation is preferred, it is not strictly required this policy will not fail if alignment is not possible. Since this policy extends the CPUManager’s Static policy, it will only apply to guaranteed pods with whole CPU requests.

Reviewer 3: same comment about possibly replacing `scheduling` with `cpu allocation logic`. I can't remember us using "scheduling" in this context in the kubelet lingo.

Reviewer 2: updated

4. If the required amount of CPUs cannot fit within an uncore cache group and there are enough schedulable CPUs on the node, assign cores in numerical order.
5. Container will not be scheduled on the node if there are not enough CPUs to satisfy the container requirements. 

In the case where the NUMA boundary is larger than a socket (setting NPS0 on a dual-socket system), the allocatable pool of CPUs does not exapnd beyond the respective socket when NUMA hints are provided in the above scheduling policy. The filtered subset of CPUs will still remain within the socket. 

In the case where the NUMA boundary is smaller than an uncore cache (enabling Sub-NUMA clustering on a monolithic cache system), the allocatable pool of CPUs does not expand beyong the respective NUMA boundary when uncore cache hints are provided in the above scheduling policy. The filtered subset of CPUs will still remain within the NUMA boundary.

> This scheduling policy will minimize the distribution of containers across uncore caches to improve performance while still maintaining the default packed logic. The scope will be initially be narrowed to implement uncore cache alignment to the default static scheduling behavior. The table below summarizes future enhancement plans to implement uncore cache alignment to be compatible with the distributed scheduling policies to reduce contention/noisy neighbor effects.

Reviewer 3: I think this explanation is much improved, thanks!
A couple allocation examples would really help clarifying the algorithm. If you have time to do that, I'd start with the case on which we need to allocate partial uncore groups both with empty groups (first allocation?) and with partially allocated groups (e.g. 1st container ever and N-th container or so)

Reviewer 2: Below the HW matrix are 5 different processors and examples of how the CPU allocation will work with containers being scheduled of various CPUs quantities and order

Reviewer 3: Fair enough, LGTM for now.

| Compatibility | alpha | beta | GA |
| --- | --- | --- | --- |
| full-pcpus-only | x | x | x |
| distribute-cpus-across-numa  |   | x | x |
| align-by-socket | x | x | x |
| distribute-cpus-across-cores |   | x | x |

### Test Plan

HW Test Matrix

> - 1P AMD EPYC 7303 32C (smt-on) NPS=1 

Reviewer 3: did we define "NPS" earlier in the document?

Reviewer 2: updated

- `/pkg/kubelet/cm/cpumanager/policy_static.go`: `2024-08-26` - `89.5%`
- `/pkg/kubelet/cm/cpumanager/topology/topology.go`: `2024-9-30` - `93.2%`

##### Integration tests

N/A. This feature requires a e2e test for testing.

##### e2e tests

- For e2e testing, checks will be added to determine if the node has a split uncore cache topology. If node does not meet the requirement to have multiple uncore caches, the added tests will be skipped. 
- e2e testing should cover the deployment of a pod that is following uncore cache alignment. CPU assignment can be determined by podresources API and programatically cross-referenced to syfs topology information to determine proper uncore cache alignment.
- For e2e testing, guaranteed pods will be deployed with various CPU size requirements on our own baremetal instances across different vendor architectures and confirming the CPU assignments to uncore cache core groupings. This feature is intended for baremetal only and not cloud instances.
- Update CI to test GCP instances of different architectures utilizing uncore cache alignment feature.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature gate flag option

> - Initial e2e tests completed and enabled

Reviewer 3: the caveat could be the tests would skip most of the time until we fix the provisioning of CI hardware (see ongoing https://github.com/kubernetes/k8s.io/issues/7339), but should not be a problem for alpha

Reviewer 5: Let's add it here explicitly, so we don't forget.

Author: Would that go here in the graduation criteria or in the caveat section?

Reviewer 2: added

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

N/A

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

> To enable this feature requires adding to the Kubelet yaml file for the CPUManager feature gate and add the policy option for uncore cache alignment

Reviewer 3: I'd just say "enable the feature gates and add the cpumanager polocy option in the kubelet configuration file"

Author: Okay ill add this basically

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `CPUManagerAlphaPolicyOptions`
  - Components depending on the feature gate: `kubelet`
- [x] Other
  - Describe the mechanism: Change the `kubelet` configuration to set a `CPUManager` policy of static then setting the policy option of `align-cpus-by-uncorecache`
  - Will enabling / disabling the feature require downtime of the control
    plane? No.
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? Yes, a `kubelet` restart is required for changes to take place.

###### Does enabling the feature change any default behavior?

No, to enable this feature, it must be explicitly set in the `CPUManager` static policy and the policy option `align-cpus-by-uncorecache` must be set.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes this feature can be disabled it will just require a restart of `kubelet`. The Kubelet configuration will need to be set with the static policy option and align-cpus-by-uncorecache flag removed.

###### What happens if we reenable the feature if it was previously rolled back?

> Feature will be enabled. Proper drain of node and restart of kubelet required. Feature is not intended to be enabled/disabled dynamically, similar to static policy.

Reviewer 3: IIRC the question is meant to assess the impacts on the running workload (expect no impact) and on the cluster infra (dependencies on control plane, need to migrate the data, API incompatibility) and both should be none

Author: Well, there could be an impact right? If I need to disable this, I will need to drain the node and sensitive workloads could be impacted

Reviewer 3: IIRC this is about the possible interdependencies between cluster components rather than the impact on the workload (cc @soltysh)

###### Are there any tests for feature enablement/disablement?

Option is not enabled dynamically. To enable/disable option, cpu_manager_state must be removed and kubelet must be restarted.
Unit tests will be implemented to test if the feature is enabled/disabled.
E2e node serial suite can be use to test the enablement/disablement of the feature since it allows the kubelet to be restarted.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

> Kubelet restarts are not expected to impact existing CPU assignments to already running workloads

Reviewer 3: The question is meant to explore how enablement of the feature in a fleet can fail, or worse its rollback. I think it's safe to say we expect no consequences if the node are properly drained and restarted, both in rollout and rollback; there are known cpumanager issues (require deletion of the state file) which predate this work. The workload is also expected to be affected.

Author: When we say rollback I usually envision moving back a version, in this case we don't have that. Would a rollback be that we just turned it off?

Reviewer 3: CC @soltysh I don't remember if we mean feature rollback, version rollback or both ( :) )

###### What specific metrics should inform a rollback?

> N/A

Reviewer 3: the reason being the alignment is best-effort, so it cannot cause additional failures

Reviewer 5: I'd suggest adding increased  pod startup latency as an information for a rollback, at minimum.

Reviewer 3: yes, that would make sense

Author: I'm not very clear on this, I get the pod startup latency but I'm not understanding how this relates to a rollback, since a rollback would require draining the node

Reviewer 3: Cluster administrator do monitor their cluster after they upgraded, and they need a way to learn if the new features they deployed are working or not; if not, they may want to rollback their cluster (or, hopefully, their test nodes on green/blue deployments) to the previous state. Which brings the question: how should cluster admins learn if the feature is working, if the cluster is working, if the affected nodes are working, thus they may need a rollback? Hence the PRR question.

Example: pod admission time (startup time) "too high" -> node misbehaving -> rollback to safe state to keep serving production traffic

Reviewer 2: added

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> N/A

Reviewer 3: because the feature is self-contained and have no dependencies outside the kubelet

Author: Added

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

N/A

### Monitoring Requirements

> Include CacheID info in podresources API to be able to verify assignment.

Reviewer 3: I think this is unnecessary? we can infer the cacheid (navigating the sysfs) through the cpuid which is already reported by podresources API

Reviewer 2: Agreed. Updated section

###### How can an operator determine if the feature is in use by workloads?

> Reference podresources API to determine CPU assignment and CacheID assignment.

Reviewer 3: and adding new metrics: https://github.com/kubernetes/kubernetes/pull/127155 like alignement by uncore/l3 group

Reviewer 2: added container_aligned_compute_resource_count metric

###### How can someone using this feature know that it is working for their instance?

> Reference podresources API to determine CPU assignment and CacheID assignment.

Reviewer 3: at workload level this is till the best option because the metrics aren't per-pod to avoid excessive cardinality

<!--
For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
for each individual pod.
Pick one more of these and delete the rest.
Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 

> - [ ] Other (treat as last resort)

Reviewer 3: please select this and move the above content from line 531 here

Reviewer 2: updated

  - Details:

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

Measure the time to deploy pods under default settings and compare to the time to deploy pods with align-by-uncorecache enabled. Time difference should be negligible.

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

- Metrics
  - Pod deployment time (seconds)

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

> N/A

Reviewer 3: I'd mention https://github.com/kubernetes/kubernetes/pull/127155 and its extension to uncore/l3 grouping

Reviewer 2: updated

implementation difficulties, etc.).
-->

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

No

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

All of the housekeeping for this feature is node internal, and thus will not require the kubelet request anything new of the apiserver

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

NA. 

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

> Pod startup time can directly affected because CPUManager will have to do a few extra steps when scheduling a Pod.

Reviewer 3: ...but we expect the extra time to be negligible as all the computation is done on RAM

Author: I'm going to just add this as additional sentence

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

> N/A

Reviewer 3: no consequence, let's maybe mention it explicitely

Author: Something like this look good?

There is no known consequence since this component is dedicated for CPU allocation for Pods which does not directly interact with any API server.

###### What are other known failure modes?

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.

> -->

Reviewer 3: since the feature is best effort, the known failure mode is fail to allocate optimally; the proposed metric and the podresources API should offer enough insights.

Author: In the scenario where we can't optimally schedule would it actually be considered a failure mode? It's more like a warning, right?

Reviewer 2: updated

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded

> -->

Reviewer 3: let's mention the date of the KEP proposed

Author: Would something like this work?

- The outlined sections were filled out was created 2024-08-27.

## Drawbacks

N/A

## Alternatives

Uncore cache affinity scheduling is possible by delgating CPU allocation from the Kubelet to the container runtime and plugins. However as a consequence of using a different implementation, the topology alignment granted by the Topology Manager within the kubelet is not compatible. 
Existing Static CPU Manager can be used, but requires manual assignment and for user to only run guaranteed pods with CPU sizes matching the corresponding uncore cache CPU group of the specific node.

## Infrastructure Needed (Optional)

<!--
Use this section if you need things from the project/SIG. Examples include a
new subproject, repos requested, or GitHub details. Listing these here allows a
SIG to get the process for these resources started right away.

> -->

Reviewer 3: CI machines with cpus with split L3? we mentioned informally a few times, maybe worth to make explicit?

### Test Plan

> The different configurations can help show the above scheduling policy behavior of how containers are assigned CPUs when the prefer-align-cpus-by-uncorecache feature is enabled with contaienrs of different CPU sizes and order. These different configurations can also help outline unit and e2e test cases. Below are a few examples of the scheduling policy across different processors with different NUMA Per Socket (NPS) settings.

Reviewer 6: s/contaienrs/containers

Author: Fixed!

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

> N/A, because the feature is self-contianed and have no dependencies outside the kubelet. 

Reviewer 6: i think it's not applicable, but not because it's self-contained, but because an upgrade->downgrade->upgrade would be handled similarly as a reboot. the cpu manager state file would be removed and pods would need to be drained to be realigned on uncore cache lines in both situations.

Author: I agree with this, @ffromani do you?

Reviewer 3: Yes, seems fair and in the end we're converging on the same topic from different perspectives, so I agree

Author: Okay I added it, lmk if it's good

###### How can an operator determine if the feature is in use by workloads?

Reference podresources API to determine CPU assignment and CacheID assignment per container.
Use proposed 'container_aligned_compute_resources_count' metric which reports the count of containers getting aligned compute resources.

###### How can someone using this feature know that it is working for their instance?

Reference podresources API to determine CPU assignment.

- [x] API .status

>   - Condition name: container_aligned_compute_resource_count

Reviewer 6: technically speaking this is a metric one could use , not something in API .status field AFAICT

Reviewer 2: updated
